interchain-token-service/src/contract/execute.rs

```
@@ -0,0 +1,329 @@
+    UnknownItsAddress(Address),
+    #[error("failed to decode payload")]
+    InvalidPayload,
+}
+
+/// Executes an incoming ITS message.
+///
+/// This function handles the execution of ITS (Interchain Token Service) messages received from
+/// its sources. It verifies the source address, decodes the message, applies balance tracking,
+/// and forwards the message to the destination chain.
+pub fn execute_message(
+    deps: DepsMut,
+    cc_id: CrossChainId,
+    source_address: Address,
+    payload: HexBinary,
+) -> Result<Response, Error> {
+    ensure_its_source_address(deps.storage, &cc_id.source_chain, &source_address)?;
+
+    match ItsHubMessage::abi_decode(&payload).change_context(Error::InvalidPayload)? {
+        ItsHubMessage::SendToHub {
+            destination_chain,
+            message,
+        } => {
+            let destination_chain = to_chain_name(&destination_chain);
+            let destination_address = load_its_address(deps.storage, &destination_chain)
+                .change_context_lazy(|| Error::UnknownChain(destination_chain.clone()))?;
+
+            let destination_payload = ItsHubMessage::ReceiveFromHub {
+                source_chain: cc_id.source_chain.clone(),
+                message: message.clone(),
+            }
+            .abi_encode();
+
+            Ok(send_to_destination(
+                deps.storage,
+                deps.querier,
+                destination_chain.clone(),
+                destination_address,
+                destination_payload,
+            )?
+            .add_event(
+                ItsContractEvent::ItsMessageReceived {
+                    cc_id,
+                    destination_chain,
+                    message,
+                }
+                .into(),
+            ))
+        }
+        _ => bail!(Error::InvalidPayload),
+    }
+}
+
+fn to_chain_name(chain: &ChainNameRaw) -> ChainName {
+    ChainName::try_from(chain.as_ref()).expect("invalid chain name")
```

this one probably shouldn't be a panic right? This is being used on data supplied externally, so I think we should return an error

---

yes, no panic

---

user can't cause an error since `ChainName` just lower cases the already validated `ChainNameRaw` string. It would be a bug if validation between the two diverges.